romancal/datamodels/library.py

```diff
@@ -0,0 +1,504 @@
+        dir_path = dir_path if dir_path is not None else os.getcwd()
+
+        # output_suffix = kwargs.pop("output_suffix", None)  # FIXME this was unused
+
+        output_paths = []
+        with self:
+            for i, model in enumerate(self):
+                if len(self) == 1:
+                    index = None
+                else:
+                    index = i
+                if save_model_func is None:
+                    filename = model.meta.filename
+                    output_path, output_filename = os.path.split(path(filename, index))
+
+                    # use dir_path when provided
+                    output_path = output_path if dir_path is None else dir_path
+
+                    # create final destination (path + filename)
+                    save_path = os.path.join(output_path, output_filename)
+
+                    model.to_asdf(save_path)  # TODO save args?
+
+                    output_paths.append(save_path)
+                else:
+                    output_paths.append(save_model_func(model, idx=index))
+
+                self.discard(i, model)
+
+        return output_paths
+
+    # TODO crds_observatory, get_crds_parameters, when stpipe uses these...
+    def crds_observatory(self):
+        return "roman"
+
+    def get_crds_parameters(self):
+        raise NotImplementedError()
+
+    def finalize_result(self, step, reference_files_used):
+        with self:
+            for i, model in enumerate(self):
+                step.finalize_result(model, reference_files_used)
+                self[i] = model
+
+    def __enter__(self):
+        self._open = True
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._open = False
+        # if exc_value:
+        #     # if there is already an exception, don't worry about checking the ledger
+        #     # instead allowing the calling code to raise the original error to provide
+        #     # a more useful feedback without any chained ledger exception about
+        #     # un-returned models
+        #     return
+        # TODO we may want to change this chain to make tracebacks and pytest output
+        # easier to read.
+        if self._ledger:
+            raise BorrowError(
+                f"ModelLibrary has {len(self._ledger)} un-returned models"
+            ) from exc_value
+
+    def index(self, attribute, copy=False):
+        """
+        Access a single attribute from all models
+        """
+        # TODO we could here implement efficient accessors for
+        # certain attributes (like `meta.wcs` or `meta.wcs_info.s_region`)
+        if copy:
+            copy_func = lambda value: value.copy()  # noqa: E731
+        else:
+            copy_func = lambda value: value  # noqa: E731
+        with self:
+            for i, model in range(len(self)):
+                attr = model[attribute]
+                self.discard(i, model)
+                yield copy_func(attr)
+
```

Comment:
Random note, but IMO it would be nice to have a convenience function for splitting a ModelLibrary by exposure type here, i.e., a more general version of [this function](https://github.com/spacetelescope/jwst/blob/master/jwst/master_background/master_background_step.py#L239) in JWST's MasterBackground step.  I think slightly different versions of this already exist in multiple places in the JWST code base, and this seems like a good opportunity to make it generic.